Implement S3 snapshot manager

kemkemG0 opened the pull request:
### What I changed

-  ~I have added the `aws-sdk-s3` to the collection for managing operations on S3. I chose this SDK because it is the official one and the version 1.24.0 is well-established and stable. The implementation was based on the official sample code found at https://github.com/awslabs/aws-sdk-rust/tree/main/examples/examples/s3/src/bin.~
- Removed aws_s3_rust and replaced it with object_store. This change allows the same abstract code to be used with S3, GCS, and Azure Storage. Currently, the implementation is only for S3, but it can be easily extended to support other services.


- ~Next, I implemented the S3 functionalities in the existing `snapshots_manager.rs`. The necessary functions using `aws-sdk-s3` were implemented in `snapshots_s3_ops.rs` and called from there.~
- Deleted the previously created snapshot_s3_ops.rs and introduced a more abstract snapshot_storage_ops.rs.


- ~One challenging aspect, different from what was initially expected, involved operations like delete and download of snapshots. The process used `get_snapshot_path` for path verification followed by delete and download actions, which invariably led to errors when the paths did not exist locally on S3.~
- ~To address this, I implemented `get_s3_snapshot_path` and used a match statement to handle different scenarios.~
- ~The same changes have been made in four places: delete and download of snapshots, and delete and download of full snapshots. (Note: Additional changes were also necessary for downloading shards.)~


- Addressed a review comment by removing S3-specific functions such as get_s3_snapshot_path and get_full_s3_snapshot_path. Replaced them with more abstract functions get_snapshot_path and get_full_snapshot_path located in snapshot_storage_ops.rs to enhance abstraction.

- The use of object_store has resulted in a reduced build size.


### Test Modifications
- The test script `tests/snapshots/snapshots-recovery.sh` was modified to allow switching between `local` and `s3` based on the input arguments. For S3 tests, the config file is automatically modified using the yq command before execution.

✅  I have verified that all tests pass on my forked repository.




resolve: #4109 

/claim #4109 

### All Submissions:

* [X] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

kemkemG0 commented:
Implementation is done.

I am going to add unitests/integration test from now.

kemkemG0 commented on lib/collection/Cargo.toml:
```diff
@@ -75,6 +75,13 @@ strum = { workspace = true }
 
 tracing = { workspace = true, optional = true }
 
+# AWS
+aws-config = { version = "1.1.7", features = ["behavior-version-latest"] }
+aws-sdk-s3 = "1.24.0"
+aws-smithy-types = "1.1.8"
+aws-smithy-types-convert = { version = "0.60.8", features = ["convert-chrono"] }
```
I used the official AWS SDK for S3

generall replied:
Any particular reason for that?

kemkemG0 commented on lib/collection/src/common/snapshots_manager.rs:
```diff
@@ -9,8 +10,38 @@ use crate::common::sha_256::hash_file;
+use crate::operations::snapshot_s3_ops;
 use crate::operations::types::CollectionResult;
 
+#[derive(Clone, Deserialize, Debug, Default)]
+pub struct SnapShotsConfig {
+    pub snapshots_storage: SnapshotsStorageConfig,
+    pub s3_config: Option<S3Config>,
+}
+
+#[derive(Clone, Debug, Default)]
+pub enum SnapshotsStorageConfig {
+    #[default]
+    Local,
+    S3,
+}
+
+impl<'de> Deserialize<'de> for SnapshotsStorageConfig {
+    fn deserialize<D>(deserializer: D) -> Result<SnapshotsStorageConfig, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s: String = Deserialize::deserialize(deserializer)?;
+        match s.as_str() {
+            "local" => Ok(SnapshotsStorageConfig::Local),
+            "s3" => Ok(SnapshotsStorageConfig::S3),
+            _ => Err(serde::de::Error::custom(
+                "Invalid snapshots_storage. Use 'local' or 's3'",
+            )),
+        }
+    }
+}
+
```
These are for deserialization of config yaml

kemkemG0 commented on lib/collection/src/operations/snapshot_s3_ops.rs:
```diff
@@ -0,0 +1,239 @@
+    key: &str,
+    source_path: &str,
+) -> CollectionResult<CompleteMultipartUploadOutput> {
+    const CHUNK_SIZE: u64 = 1024 * 1024 * 5;
+    const MAX_CHUNKS: u64 = 10000;
```
I just followed official example as below.

https://github.com/awsdocs/aws-doc-sdk-examples/blob/429c855e5f8d74330f9cbc4cb1570668439283ee/rustv1/examples/s3/src/bin/s3-multipart-upload.rs#L25-L29

This probably should be on config but i wasn't sure.

kemkemG0 commented on lib/collection/src/operations/snapshot_s3_ops.rs:
```diff
@@ -0,0 +1,239 @@
+    key.map(|k| k.trim_start_matches("./").to_string())
+}
+
+pub async fn multi_part_upload(
```
I used this official example code as the reference.

https://github.com/awsdocs/aws-doc-sdk-examples/blob/429c855e5f8d74330f9cbc4cb1570668439283ee/rustv1/examples/s3/src/bin/s3-multipart-upload.rs

kemkemG0 commented on .github/workflows/integration-tests.yml:
```diff
@@ -226,10 +226,66 @@ jobs:
+
+  test-snapshot-operations-s3-minio:
+    runs-on: ubuntu-latest
+
+    # Setup minio server
+    services:
+      minio:
+        image: wktk/minio-server
+        ports:
+          - 9000:9000
+        env:
+          MINIO_ACCESS_KEY: "minioadmin"
+          MINIO_SECRET_KEY: "minioadmin"
+
```
almost same as `test-snapshot-operations-local:` but added minio server so that we can use `minio` from the shell file.
I refered https://github.com/mozilla/sccache/pull/1513 to implement this.

kemkemG0 commented on lib/collection/src/operations/snapshot_s3_ops.rs:
```diff
@@ -0,0 +1,238 @@
+    /// 3. Upload each chunk to AWS S3 using the `upload_id`. Each chunk is referred to as a part, and each part is uploaded separately.
+    /// 4. Once all parts have been uploaded, send a `complete_multipart_upload` request to AWS S3. This request will assemble all the uploaded parts into a single file based on the `upload_id`.
+    ///
+    /// This function handles each of these steps and includes error handling for each part of the process.
+    const CHUNK_SIZE: u64 = 1024 * 1024 * 50; // 50MB
+    const MAX_CHUNKS: u64 = 10000;
+
```
I wasn't sure if this should be configured from config.yml. So i'm leaving these here for now.

kemkemG0 commented on lib/collection/src/operations/shared_storage_config.rs:
```diff
@@ -46,7 +47,7 @@ impl Default for SharedStorageConfig {
             outgoing_shard_transfers_limit: DEFAULT_IO_SHARD_TRANSFER_LIMIT,
             snapshots_path: DEFAULT_SNAPSHOTS_PATH.to_string(),
-            s3_config: None,
+            snapshots_config: default::Default::default(),
```
Changed from `s3_config` to `snapshots_config` so that we can extend to other storage service

kemkemG0 commented on lib/collection/src/common/snapshots_manager.rs:
```diff
@@ -9,19 +11,51 @@ use crate::common::sha_256::hash_file;
     pub region: Option<String>,
     pub access_key: Option<String>,
     pub secret_key: Option<String>,
+    pub endpoint_url: Option<String>,
```
We need to specify URL for minIO when testing, so added this.

generall commented on lib/collection/src/collection/snapshots.rs:
```diff
@@ -306,6 +306,12 @@ impl Collection {
         Ok(absolute_snapshot_path)
     }
 
+    pub async fn get_s3_snapshot_path(&self, snapshot_name: &str) -> CollectionResult<PathBuf> {
```
This looks like an abstraction leak. Ideally only SnapshotManager should know details of the implementation

generall commented on lib/collection/src/common/snapshots_manager.rs:
```diff
@@ -32,11 +66,35 @@ pub enum SnapshotStorageManager {
+                    }
+                    if let Some(endpoint_url) = &s3_config.endpoint_url {
+                        env::set_var("AWS_ENDPOINT_URL", endpoint_url);
+                    }
```
This looks a bit artificial. Why can't we just directly pass parameters to the constructor?

generall commented on lib/collection/src/common/snapshots_manager.rs:
```diff
@@ -1,5 +1,7 @@
+use std::env;
 use std::path::Path;
 
+use aws_sdk_s3::Client as S3Client;
```
Have you considered other crates? Like e.g. https://crates.io/crates/object_store which seems to be x5 smaller and doesn't have that strong affiliation with AWS

kemkemG0 replied:
Definitely worht trying with this!

generall commented on lib/collection/src/common/snapshots_manager.rs:
```diff
@@ -9,19 +11,51 @@ use crate::common::sha_256::hash_file;
+                "Invalid snapshots_storage. Use 'local' or 's3'",
+            )),
+        }
+    }
```
Why custom deserializer is required in this case?

If it is required, how to ensure that we won't forget to add a support for a new type here if enum is changed?

generall commented on lib/storage/src/content_manager/snapshots/mod.rs:
```diff
@@ -54,15 +55,27 @@ pub async fn get_full_snapshot_path(
+    let snapshot_manager = toc.get_snapshots_storage_manager().await;
+    let snapshot_dir = match snapshot_manager {
+        SnapshotStorageManager::LocalFS(_) => get_full_snapshot_path(toc, snapshot_name).await?,
+        SnapshotStorageManager::S3(_) => get_full_s3_snapshot_path(toc, snapshot_name).await?,
```
`snapshot_manager` should have enough info to resolve this  internally.

kemkemG0 replied:
I initially considered this, but `get_full_snapshot_path` needs to be relocated under `snapshot_manager` to achieve this so I didn't do that. But yeah, I will definitely address this. : )

Perhaps I'll create a separate PR for this fix, as the current one is already quite large.

generall commented on src/actix/api/snapshot_api.rs:
```diff
@@ -70,8 +71,31 @@ pub async fn do_get_full_snapshot(
+            let snapshot_path = get_full_snapshot_path(toc, snapshot_name).await?;
+            Ok(NamedFile::open(snapshot_path)?)
+        }
+        SnapshotStorageManager::S3(_) => {
```
As well, should be implementation-specific. Exposing internal logic of the snapshot manager into user code means we are breaking abstraction

generall commented on src/common/snapshots.rs:
```diff
@@ -65,10 +66,19 @@ pub async fn delete_shard_snapshot(
+                .get_shard_snapshot_path(shard_id, &snapshot_name)
+                .await?
+        }
+        SnapshotStorageManager::S3(_) => {
```
same, should be performed in `snapshot_manager` internally

generall reviewed (changes requested):
Hey @kemkemG0, thanks for the PR!

Please let me know if you are filling to take care of the comments or if you need help with those

kemkemG0 commented on lib/collection/src/collection/mod.rs:
```diff
@@ -667,6 +667,14 @@ impl Collection {
     pub fn shared_storage_config(&self) -> Arc<SharedStorageConfig> {
         self.shared_storage_config.clone()
     }
+
+    pub fn snapshots_path(&self) -> &Path {
+        &self.snapshots_path
+    }
+
+    pub fn shards_holder(&self) -> Arc<LockedShardHolder> {
+        self.shards_holder.clone()
+    }
```
These were needed from `get_snapshot_path` , `get_full_snapshot_path` and `get_shard_snapshot_path`

kemkemG0 commented on lib/collection/src/common/snapshots_manager.rs:
```diff
@@ -9,34 +11,83 @@ use crate::common::sha_256::hash_file;
 #[allow(dead_code)]
-pub struct SnapshotStorageS3 {
-    s3_config: S3Config,
+pub struct SnapshotStorageCloud {
+    client: Box<dyn object_store::ObjectStore>,
```
not only s3 nor gcs but an abstraction for cloud storage,

kemkemG0 commented on lib/collection/src/common/snapshots_manager.rs:
```diff
@@ -9,34 +11,83 @@ use crate::common::sha_256::hash_file;
-            SnapshotStorageManager::S3(SnapshotStorageS3 { s3_config })
-        } else {
-            SnapshotStorageManager::LocalFS(SnapshotStorageLocalFS)
+    pub fn new(snapshots_config: SnapShotsConfig) -> CollectionResult<Self> {
+        match snapshots_config.clone().snapshots_storage {
+            SnapshotsStorageConfig::Local => {
+                Ok(SnapshotStorageManager::LocalFS(SnapshotStorageLocalFS))
+            }
+            SnapshotsStorageConfig::S3 => {
+                let mut builder = AmazonS3Builder::new();
+                if let Some(s3_config) = &snapshots_config.s3_config {
```
This is the only place where abstraction is broken.
builder's here.

kemkemG0 commented:
Hi @generall , thanks for the review!!

I fix them up! 


- Removed `aws_s3_rust` and replaced it with `object_store`. This change allows the same abstract code to be used with S3, GCS, and Azure Storage. Currently, the implementation is only for S3, but it can be easily extended to support other services.

- Deleted the previously created `snapshot_s3_ops.rs` and introduced a more abstract `snapshot_storage_ops.rs`.

- Addressed a review comment by removing S3-specific functions such as `get_s3_snapshot_path` and `get_full_s3_snapshot_path`. Replaced them with more abstract functions `get_snapshot_path` and `get_full_snapshot_path` located in `snapshot_storage_ops.rs` to enhance abstraction.

- The use of `object_store` has resulted in a reduced build size.

kemkemG0 commented on lib/collection/src/common/snapshots_manager.rs:
```diff
@@ -9,34 +11,83 @@ use crate::common::sha_256::hash_file;
 pub enum SnapshotStorageManager {
     LocalFS(SnapshotStorageLocalFS),
-    S3(SnapshotStorageS3),
+    // Assuming that we can have common operations for all cloud storages
+    S3(SnapshotStorageCloud),
+    // <TODO> : Implement other cloud storage
+    // GCS(SnapshotStorageCloud),
+    // AZURE(SnapshotStorageCloud),
```
Assuming that we can use `SnapshotStorageCloud` for all.
If it's not possible, we need to switch back to implementing`SnapshotStorageS3` or `SnapshotStorageGCS` for each.

kemkemG0 commented on lib/collection/src/common/snapshots_manager.rs:
```diff
@@ -187,33 +321,164 @@ impl SnapshotStorageLocalFS {
+    /// Get absolute file path for a full snapshot by name
+    ///
+    /// This enforces the file to be inside the snapshots directory
+    async fn get_full_snapshot_path(
+        &self,
+        snapshots_path: &str,
+        snapshot_name: &str,
+    ) -> CollectionResult<PathBuf> {
+        let absolute_snapshot_dir = Path::new(snapshots_path).canonicalize().map_err(|_| {
+            CollectionError::not_found(format!("Snapshot directory: {snapshots_path}"))
+        })?;
+
+        let absolute_snapshot_path = absolute_snapshot_dir
+            .join(snapshot_name)
+            .canonicalize()
+            .map_err(|_| CollectionError::not_found(format!("Snapshot {snapshot_name}")))?;
+
+        if !absolute_snapshot_path.starts_with(absolute_snapshot_dir) {
+            return Err(CollectionError::not_found(format!(
+                "Snapshot {snapshot_name}"
+            )));
+        }
+
+        if !absolute_snapshot_path.is_file() {
+            return Err(CollectionError::not_found(format!(
+                "Snapshot {snapshot_name}"
+            )));
+        }
+
+        Ok(absolute_snapshot_path)
```
Moved from `lib/storage/src/content_manager/snapshots/mod.rs`

kemkemG0 commented on lib/collection/src/common/snapshots_manager.rs:
```diff
@@ -187,33 +321,164 @@ impl SnapshotStorageLocalFS {
+    async fn get_snapshot_path(
         &self,
-        _directory: &Path,
-    ) -> CollectionResult<Vec<SnapshotDescription>> {
-        unimplemented!()
+        snapshots_path: &Path,
+        snapshot_name: &str,
+    ) -> CollectionResult<PathBuf> {
+        let absolute_snapshot_dir = snapshots_path.canonicalize().map_err(|_| {
+            CollectionError::not_found(format!("Snapshot directory: {}", snapshots_path.display()))
+        })?;
+
+        let absolute_snapshot_path = absolute_snapshot_dir
+            .join(snapshot_name)
+            .canonicalize()
+            .map_err(|_| CollectionError::not_found(format!("Snapshot {snapshot_name}")))?;
+
+        if !absolute_snapshot_path.starts_with(absolute_snapshot_dir) {
+            return Err(CollectionError::not_found(format!(
+                "Snapshot {snapshot_name}"
+            )));
+        }
+
+        if !absolute_snapshot_path.is_file() {
+            return Err(CollectionError::not_found(format!(
+                "Snapshot {snapshot_name}"
+            )));
+        }
+
+        Ok(absolute_snapshot_path)
+    }
```
moved from `lib/collection/src/collection/snapshots.rs`

generall commented:
Hey @kemkemG0, nice job! I think it is finally coming to the desired shape!

/tip $200 @kemkemG0

generall commented:
A few things we would need to finalize this PR:

- In the `do_get_snapshot` and similar functions we download the snapshot from s3 and only after it we respond to te user. But this might not work so good with large files, where user would need to wait indefinitely without a single response byte. Additionally it will left temporary files behind which are never cleaned. 

My suggestion is to stream files from either local or S3 as an 

```
HttpResponse::Ok().content_type("application/octet-stream").streaming(stream)
```

In order to do that, we would need to implement something like

```
pub struct SnapshotStreamer {
    snapshot_manager: SnapshotStorageManager,
    snapshot_path: PathBuf,
}

impl Stream for SnapshotStreamer {
    type Item = Result<bytes::Bytes, CollectionError>;
```

`NamedFile` from actix does a lot of nice things, maybe we could either keep it for local files, or re-implement in the octet-stream.

- make sure tests work

Please note that the main issue bounty should be paid once this PR is merged

kemkemG0 commented:
@generall 
I also noticed that, shard snapshots had the separated implementations and need to add `create_shard_snapshot`, `recover_shard_snapshot`, `create_shard_snapshot` and `delete_shard_snapshot` under `snapshot_storage_manager`.

These looks complicated and may take some time.

kemkemG0 commented:
- [ ] Fixed bugs and tests (should) passe
- [ ] Implement Stream later on

kemkemG0 commented on .github/workflows/integration-tests.yml:
```diff
@@ -254,3 +310,35 @@ jobs:
           sleep 10
           ./tests/shard-snapshot-api.sh test-all
+
+  test-shard-snapshot-api-s3-minio:
```
Add Integration test for s3 version.

kemkemG0 commented on lib/collection/src/common/snapshot_stream.rs:
```diff
@@ -0,0 +1,42 @@
+use actix_files::NamedFile;
+use actix_web::{HttpRequest, HttpResponse, Responder};
+use futures::Stream;
+
+pub struct SnapShotStreamLocalFS {
+    pub snapshot_path: PathBuf,
+    pub req: HttpRequest,
+}
+pub struct SnapShotStreamCloudStrage {
+    pub streamer:
+        std::pin::Pin<Box<dyn Stream<Item = Result<bytes::Bytes, object_store::Error>> + Send>>,
+}
+
+pub enum SnapshotStream {
+    LocalFS(SnapShotStreamLocalFS),
+    CloudStorage(SnapShotStreamCloudStrage),
+}
+
+impl Responder for SnapshotStream {
+    type Body = actix_web::body::BoxBody;
+
+    fn respond_to(self, _: &actix_web::HttpRequest) -> HttpResponse<Self::Body> {
+        match self {
+            SnapshotStream::LocalFS(stream) => match NamedFile::open(stream.snapshot_path) {
+                Ok(file) => file.into_response(&stream.req),
+                Err(e) => match e.kind() {
+                    std::io::ErrorKind::NotFound => {
+                        HttpResponse::NotFound().body(format!("File not found: {}", e))
+                    }
+                    _ => HttpResponse::InternalServerError()
+                        .body(format!("Failed to open file: {}", e)),
+                },
+            },
+
+            SnapshotStream::CloudStorage(stream) => HttpResponse::Ok()
+                .content_type("application/octet-stream")
+                .streaming(stream.streamer),
+        }
+    }
+}
```
Add `SnspShotStream` for downloading snapshot with streaming

kemkemG0 commented:
@generall 
I fixed them up and added extra integration test !
